### ISSUE TITLE
Add more diagnostic methods and tests for bayesnecfit objects

### ISSUE THREAD

**Issue by beckyfisher** [open-AIMS/bayesnec#148]
At the moment most diagnostic tests only check for model convergence. Give the sensitivity of NSEC on estimates of variance, diagnostic tests to ensure that the model simulated variance is representative of the input data would provide surety that the estimates are valid.

**Comment by beckyfisher** [comment]
IN the first instance we could implement a pp_check method for the bayesnecfit class.
This include some examples in the single model usage vignette and/or it may be worth building a dedicated model checking vignette.

**Comment by beckyfisher** [comment]
## Scoping this out

The issue as written asks for "diagnostic tests to ensure that the model simulated variance is representative of the input data". That is the right target, but it is not yet a specification, so this comment tries to turn it into one: what exists, where the actual hole is, what would fill it, and what has to be decided before any of it can be written.

Two things are settled already and are assumed here rather than re-argued. First, the split recorded in D6: sensitivity to the *estimator's* own choices (`sig_val`, the reference definition, `resolution`) belongs to `toxval`, because it applies equally to `nsec.drc` and `nsec.brmsfit`; sensitivity to what the *fit* supplies belongs here. Nothing below re-opens that. Second, `pp_check` was already proposed in the comment above, and it is part of the answer — but only part, and not the interesting part.

---

### 1. What a `bayesnec` fit currently offers as a diagnostic

| | what it checks |
|---|---|
| `check_chains()` | mixing, autocorrelation |
| `rhat()` | convergence |
| `check_priors()` | prior vs posterior overlap |
| `loo` / `waic` in `mod_stats` | relative predictive fit between candidate models |
| `dispersion()` | observed vs simulated Pearson SS — **Poisson and Binomial only** |
| `plot()` / `autoplot()` | the curve, by eye |

Everything in that list except `dispersion()` is about the sampler or about *relative* model ranking. Nothing asks whether the fitted model's implied variability matches the data's — which is what the issue is about, and the one assumption that `nsec` rests on most heavily.

### 2. Why this lands on NSEC specifically

`nsec` sets its reference at `R/nsec.R:155`:

```r
reference <- quantile(p_samples[, 1], sig_val)
```

`p_samples` is `posterior_epred`, so this is the `sig_val` quantile of the **posterior of the control mean**, and the width of that posterior is set by the fitted dispersion parameter at the control. Over-state control variability and the reference falls, the curve crosses it later, and `nsec` moves *right* — a less protective number. Under-state it and `nsec` moves left. The bias is not conservative in a known direction, and nothing in the current output tells a user which way it has gone.

This is not hypothetical. D6 already records it on the `example1` simulation: `nsec` moved from 0.93 [0.43, 1.32] to 1.09 [0.50, 1.53] under `disp("power")` while `ecx` *tightened*. The estimator was identical; only the variance model changed. The `example1` vignette states the consequence outright — "an `nsec` reported from a constant-dispersion fit rests on an assumption about control variability that the fit has not checked". #148 is the issue for checking it.

`ecx` is comparatively insulated: it is read off the fitted curve relative to its own asymptotes, so a variance error moves its interval but not its location.

### 3. Why the existing `dispersion()` cannot be the answer

Two independent reasons, and the second is the one that shapes the design.

**(a) Family coverage.** `R/dispersion.R:53` restricts it to `poisson` and `binomial`. Eight of the ten supported families — `gaussian`, `Gamma`, `negbinomial`, `beta_binomial`, `Beta`, `bernoulli`, and the two-block families — get `numeric()` back and an `NA` in `mod_stats`. So the families most used for continuous ecotox endpoints have no variance diagnostic at all.

**(b) A global statistic is uninformative for any family with a free dispersion parameter.** `gaussian`, `Gamma`, `negbinomial`, `Beta` and `beta_binomial` all estimate a dispersion parameter, and that parameter absorbs exactly the discrepancy the Pearson ratio measures. On the packaged `manec_example` (gaussian, `nec4param`) I get a global ratio of **1.011 [0.71, 1.44]** — a clean bill of health.

The same fit, with the ratio computed on residuals within six quantile bins of `x`:

| bin of `x` | 0.03–0.34 | 0.34–0.61 | 0.61–0.88 | 0.88–1.36 | 1.36–1.77 | 1.77–3.22 |
|---|---|---|---|---|---|---|
| observed resid SD | 0.412 | 0.292 | 0.430 | 0.548 | 0.320 | 0.876 |
| simulated resid SD | 0.525 | 0.510 | 0.503 | 0.508 | 0.508 | 0.525 |
| **ratio** | **0.79** | 0.57 | 0.86 | 1.08 | 0.63 | **1.67** |

The constant-`sigma` model simulates a flat ~0.51 everywhere. The data do not. In the control region the model simulates **27% more variability than the data show**, which is precisely the quantity `nsec` keys off — and the global statistic reported it as 1.01. (17 observations per bin, so treat the individual numbers as indicative; the point is that the global summary is structurally incapable of showing this, not that these particular bins are significant.)

**So the diagnostic has to be local, not global.** That is the core finding of this scoping exercise, and it is what separates #148 from "add `pp_check`".

---

### 4. Proposed scope

Three parts, separable, in increasing order of cost. Part B is the actual issue; A and C are the framing around it.

#### A. `pp_check()` methods — cheap, and already 80% free

`pp_check(pull_brmsfit(fit))` works today. What is missing is dispatch, so that a user does not have to know to unwrap. For `bayesnecfit` it is a one-line delegation. `brms` is in `Depends`, so `importFrom(brms, pp_check)` adds no dependency.

The interesting case is `bayesmanecfit`, and it is a genuine fork — see decision 2 below.

For `bayesnechurdlefit` there is no combined posterior predictive to check: the two components carry different response vectors (growth is fitted on survivors only, survival on `.alive`). It would follow `dispersion()`'s existing precedent and return one result per component, or take `which =`.

**Cost:** small for `bayesnecfit` + hurdle; the `bayesmanecfit` decision drives the rest.

#### B. A local variance check — the substance of #148

A function that reports, per concentration group, the observed variability against what the fitted model simulates, with a posterior predictive p-value. Sketch:

```r
check_variance(fit, group = NULL, stat = "sd")
#>   group     n  obs_sd  sim_sd  ratio   ppp
#> 1 0.1      12   0.081   0.142  0.571 0.008
#> 2 0.3      12   0.095   0.141  0.674 0.031
#> ...
```

with a `plot`/`autoplot` method, and an explicit flag on the control row, since that is the row `nsec` depends on. Naming: `check_variance()` sits with the existing `check_chains()` / `check_priors()` / `check_formula()` family, and avoids the confusion of a `check_dispersion()` sitting next to `dispersion()`.

Two design constraints worth surfacing now:

- **Grouping needs replication.** `herbicide` has 12 per concentration, `nassarius` 18–36 — fine. But `nec_data` has 100 unique `x` values and no replication at all, which is also what a continuous covariate looks like in practice. Binning is then required, and the bin width is a user-visible choice.
- **The statistic must be residual-based, not raw.** Within a bin, the raw SD of `y` mixes residual variability with the slope of the curve across the bin. On `manec_example` the top bin's raw SD is 1.72 against a residual SD of 0.88 — the difference is entirely curve, not noise. Getting this wrong makes every steep-region bin look over-dispersed.

The natural side-effect is to **generalise `dispersion()` past Poisson/Binomial** while keeping the global statistic where it is genuinely informative (the two families with no free dispersion parameter) and pointing users at the local check otherwise. That in turn is what #149 (automatic family switching on over-dispersion) would need to consume, so it is worth building with that in mind even if #149 stays deferred.

**Cost:** the real work. New file, new S3 generic across three classes, tests, docs.

#### C. Quantile residuals — where #56 belongs

You are right that #56 is the same territory. I would argue it is a *sub-part* of #148 rather than an alternative to it, and that the `DHARMa` question has a cheaper answer than it did when #56 was raised.

`DHARMa`'s core output is a uniform quantile residual, and it would be built here by handing `createDHARMa()` a `posterior_predict()` matrix — about 15 lines. But the Bayesian equivalent already exists inside the current dependency set: **LOO-PIT**, via `pp_check(fit, type = "loo_pit_overlay")`, needing only `add_criterion(fit, "loo")`, which `bnec()` already does at `R/add_criteria.R`. `loo` is in `Imports`.

So the choice is between a new `Suggests` dependency and a set of plots users already know, versus zero new dependencies, a properly Bayesian residual, and documentation work to explain it. `DHARMa`'s tests are also frequentist in framing (a p-value against a null of correct specification), which sits awkwardly next to everything else the package reports.

**Recommendation: close #56 into #148**, deliver LOO-PIT through the `pp_check` methods from part A, and document the recipe. Revisit `DHARMa` only if users ask for it by name.

---

### 5. What I would leave out

- Anything about `sig_val`, the reference definition, or `resolution` — D6 puts that in `toxval`.
- Prior sensitivity. D6 lists it under the `bayesnec` half, and it belongs there, but it is a different mechanism (prior-data conflict, not variance misspecification) and a different tool. Better as its own issue than bolted onto this one.
- Automatic remediation. Detecting a variance problem and *fixing* it — switching family, adding `disp()` — is #149, and should stay separate.

### 6. Rough size

Part A alone is a small PR. Parts A + B + C together are a substantial one: a new exported generic across three classes, `dispersion()` generalised, and either a new model-checking vignette or a new section in `example1`. Note the vignette side is gated on #190 — anything added to `example1.Rmd.orig` will not render until the full `precompile.R` run.

---

### Decisions needed

1. **Is part B in scope, or is #148 just `pp_check`?** The comment above proposes `pp_check` "in the first instance". If that is the whole of it, this closes cheaply and the variance question stays open. My argument is that `pp_check` alone does not answer the issue as written — the demonstration in §3 is a fit that passes every global check and still mis-states control variability by 27%.

2. **`pp_check` for `bayesmanecfit` — which of three?** (a) facet by candidate model, honest but never shows the averaged model the user is actually reporting; (b) a weighted-mixture `yrep` drawn from the model-averaged posterior predictive, which is the right object but needs `bayesplot` in `Imports` to render (it ships with `brms` so it is free at install time, but it is formally a new dependency); (c) build the averaged check in `ggplot2` directly, which is already in `Depends`, at the cost of writing the plotting code. Note that (b) and (c) both depend on PR #217 having landed, or the averaged `yrep` is redrawn on every call.

3. **Grouping in `check_variance()` — require replication, or bin?** Requiring genuine replicates is defensible and matches how these designs are run, but it refuses `nec_data`, the package's own example. Automatic binning always returns something, including where the answer is meaningless. Third option: require replication by default, allow explicit `group = <bin spec>` as an override.

4. **#56 — fold in, or keep separate?** My recommendation above is fold in and prefer LOO-PIT over `DHARMa`. This is the dependency call, so it is yours.

5. **Which classes must be covered before this closes?** `bayesnecfit` is straightforward. `bayesmanecfit` is decision 2. `bayesnechurdlefit` and the two-block families (`hurdle_gamma`, `zero_inflated_beta`, and the two zero-inflated count families) each need a decision about what "the simulated variance" even means for a mixture — the same question that made `disp()` refuse them (`R/disp_model.R`). It would be reasonable to scope this issue to non-mixture families and raise the mixture case separately.

6. **Numeric test, or plot?** The issue says "diagnostic tests". `check_chains()` is a plot; `dispersion()` is a number. A table plus a `plot` method covers both, but it is worth confirming that a returned object with a posterior predictive p-value per group is what is wanted, rather than a plot to read by eye.

Decisions 1 and 2 gate everything else. 3–6 can be settled during implementation if 1 and 2 are answered.

**Comment by beckyfisher** [comment]
## Decisions settled

| | |
|---|---|
| 1 | Part B **in scope**, and extended — see *Control lack-of-fit* below |
| 2 | Averaged check built in `ggplot2` directly. No `bayesplot` dependency |
| 3 | Require replication; **fall back to automatic binning with a warning** |
| 4 | #56 folds into #148. LOO-PIT via `pp_check`, no `DHARMa` |
| 6 | Both a numeric test and a plot |

Decision 5 needed an answer rather than a choice, so it is below.

---

### Decision 5 — the two-block families are fine, and I was wrong to flag them

I raised the mixture families as a problem by analogy with `disp()`, which refuses them (`R/disp_model.R`). That analogy does not hold, and checking `brms` settles it. `posterior_predict` for all four draws the **full mixture, including the point mass at zero**:

```r
# brms:::posterior_predict_hurdle_gamma
ifelse(tmp < hu, 0, rgamma(ndraws, shape = shape, scale = scale))
# brms:::posterior_predict_zero_inflated_beta
ifelse(tmp < zi, 0, rbeta(ndraws, shape1 = mu * phi, shape2 = (1 - mu) * phi))
```

`zero_inflated_poisson` and `zero_inflated_negbinomial` are the same shape. So `yrep` is a draw of the response *as observed*, and every statistic — density, grouped SD, control mean — is well defined with no decomposition. Nothing special is needed.

The reason `disp()` refuses these families was never about checking. It is that a variance function placed on the *component's* dispersion parameter does not describe the dispersion of the *response*, which is a mixture. That is a modelling problem. A diagnostic has no such problem because `posterior_predict` hands you the mixture itself.

There is in fact a statistic here that exists **only** for these families and is worth reporting: `mean(y == 0)` against `mean(yrep == 0)`. Whether the zero fraction is right is the whole question these families were added for (#104), and nothing currently reports it.

**`bayesnechurdlefit` — you are right, per-component is straightforward.** One row set per component, exactly as `dispersion()` already does. Worth noting there is a cheap third option on top: because the two posteriors are independent and the class already pairs draws row-wise for `hurdle_component_preds()`, a **combined** check can be simulated directly — draw alive/dead from the survival fit, then a growth value for each survivor — which reconstructs the full observed response including its zeros. That is the only check of the hurdle fit *as a model of the data the user handed in*, rather than of its two halves separately. Small, and I would include it, but flagging it rather than assuming (new decision (d) below).

---

### Control lack-of-fit — decision 1's addition

This is a better fit with the issue than I think you were claiming for it. It is not a second feature bolted on; it is the other half of the same failure.

`nsec` sets its reference from `p_samples[, 1]`. With the default `x_range = NA`, `bnec_newdata()` builds the grid as `seq(min(x_vec), max(x_vec), ...)` (`R/bnec_newdata.R:48`), so column 1 is the **fitted curve at `min(x)`**. The observed control is `y[x == min(x)]` — already the package's convention, at `R/check_data.R:76`. Same rows. So:

- a fit that **over- or under-states control variability** biases the reference in *spread* (my §2 above);
- a fit whose curve is **pulled away from the control data** biases the reference in *location*.

Both land on the same number, and neither is currently reported. The location half is arguably worse, because it is a bias in the point estimate rather than in the interval.

The statistic is the natural partner of the variance one — a posterior predictive check on the control mean, rather than on the control SD:

```
obs_mean(control)  vs  distribution of mean(yrep[, control])   -> ratio, ppp
obs_sd(control)    vs  distribution of sd(yrep[, control])     -> ratio, ppp
```

So part B becomes one table with two statistics per group, location and scale, and the control row flagged. Which suggests the name should be `check_fit()` rather than `check_variance()` — it sits as a coherent third alongside `check_chains()` (the sampler) and `check_priors()` (the priors): the fit against the data.

**One consequence worth being explicit about: model averaging does not protect you here.** Stacking weights come from a global `elpd`. The control is a handful of rows out of many, so a candidate model can hold high weight while fitting the control badly — it wins on the bulk of the curve and pays almost nothing for the control. That is exactly the case you describe, and it means the table has to report **per-candidate-model rows as well as the averaged row** for a `bayesmanecfit`, or it cannot tell you which model is doing the damage. I have assumed that rather than raised it as a decision; say if you disagree.

---

### Revised specification

**A. `pp_check()` methods.** `importFrom(brms, pp_check)` — `brms` is in `Depends`, no new dependency. `bayesnecfit` delegates to `pull_brmsfit()`. `bayesmanecfit` builds the model-averaged `yrep` and renders in `ggplot2` (decision 2); depends on PR #217 for a stable draw. `bayesnechurdlefit` returns per component, plus the combined simulation if (d) is yes. Includes documenting the LOO-PIT recipe (decision 4) — `bnec()` already stores the `loo` criterion at `R/add_criteria.R`, so `type = "loo_pit_overlay"` needs nothing added.

**B. `check_fit()`.** A table, one row per group, with:

| | |
|---|---|
| `obs_mean`, `sim_mean`, `ratio`, `ppp` | location — does the curve go through the data here |
| `obs_sd`, `sim_sd`, `ratio`, `ppp` | scale — is the simulated variance representative |
| `p_zero_obs`, `p_zero_sim` | two-block families only |
| `control` | flag on the `x == min(x)` row |

Grouping: replicate levels of `x` where they exist; automatic binning with a warning where they do not (decision 3). `check_data.R:76` already uses `length(ctl) >= 3` as its usability threshold, so that is the natural trigger. Statistics are computed on residuals from the fitted mean, not on raw `y`, for the reason in §3 above. Plus `plot`/`autoplot` methods (decision 6). Generalises `dispersion()` past Poisson/Binomial as a side effect, keeping the global Pearson statistic where it is informative — the two families with no free dispersion parameter — and pointing at the local table otherwise.

*Caveat to document:* a posterior predictive p-value uses the data twice and is therefore conservative — it under-detects. It is a flag, not a test. LOO-PIT from part A is the calibrated companion, which is a further argument for delivering them together.

**C.** Folded into A.

---

### New decisions arising

**(a) Name.** `check_fit()` covering both statistics, or split into `check_variance()` and a separate control check? I prefer one function — the two statistics are read together and share all the grouping machinery.

**(b) Does anything surface automatically?** A control lack-of-fit is the kind of thing a user will not go looking for. Options: on demand only; a line in `summary()`; or a message at the end of `bnec()`, which has precedent in `check_data()`'s normalisation messages. This changes what users see by default, so it is yours.

**(c) Does `check_fit()` supersede `dispersion()` or sit beside it?** Beside it is safer — `dispersion()` is documented in `example1` and appears in `mod_stats` — but then two functions report overlapping things and we should say in the docs when each applies.

**(d) The combined hurdle simulation** described above: in or out?

---

### Size and sequencing

Larger than the original issue implies, but coherent: one new exported generic across three classes with `plot`/`autoplot`, `pp_check` methods, `dispersion()` generalised, tests, and either a new model-checking vignette or a section in `example1`. The vignette side is gated on #190 — anything added to `example1.Rmd.orig` will not render until the next full `precompile.R` run.

Part A is separable and could land first as a small PR. Part B is the substance and wants the `bayesmanecfit` averaging from PR #217 already merged.

With (a)–(d) answered this is specific enough to leave the "needs a decision first" list.

**Comment by beckyfisher** [comment]
## Final decisions

| | |
|---|---|
| (a) | `check_fit()` — one function covering both statistics |
| (b) | message at the end of `bnec()` **and** a line in `summary()` |
| (c) | keep `dispersion()` as well |
| (d) | combined hurdle simulation in |

That closes the scoping. Full spec written up in `notes/tasks/148-model-fit-diagnostics.md`.

Two things came out of verifying the spec against a live fit that change the shape of (b), so recording them here rather than only in the note.

### LOO-PIT recomputes `loo` on every call — keep it on demand

`brms:::pp_check.brmsfit` calls, unconditionally, in the `loo_pit*` path:

```r
loo_object <- do_call(loo, c(pred_args, save_psis = TRUE))
```

It ignores the stored criterion entirely. Passing `save_psis = TRUE` to `add_criterion()` does not help — on 2.23.0 no `psis_object` is retained either way, and the recompute happens regardless. On a `bayesmanecfit` that is a fresh `loo` per candidate model, per call.

So LOO-PIT stays on demand. **The end-of-`bnec()` message must use the posterior-predictive statistic**, which is cheap — one `posterior_predict` at the observed data, which is a rounding error next to the fit itself. That is fine, because the control check is posterior-predictive anyway; nothing in the automatic path needs `loo`.

### The control check has little power when the control has few replicates

On a live `hurdle_gamma` fit with 6 control observations:

| | obs | sim | ppp |
|---|---|---|---|
| control mean | 4.50 | 5.53 | 0.82 |
| control sd | 1.53 | 2.58 | 0.82 |

Both statistics compute cleanly on a two-block family with no special-casing, which is the thing I wanted to confirm. But neither `ppp` is anywhere near flagging, on a fit where the simulated control mean is 23% above the observed one. With n = 6 there is simply not much to test against.

That is the same constraint driving decision 3's binning fallback, and it matters more for (b) than for anything else: an automatic message that stays silent on a 23% control discrepancy is worse than no message, because silence reads as a pass. Two options, and I would take the second:

1. Threshold on the `ppp` alone — simple, and will rarely fire on the small control groups that are common in these designs.
2. Threshold on the **ratio** with the `ppp` reported alongside — fires on effect size, which is what a user can act on, and does not go quiet just because the control is small. The message would then say what the discrepancy is rather than asserting significance.

Also folded into the note: the check must **not** be cached on the fitted object — #180 / PR #205 went the other way, removing `pred_vals$posterior` precisely to stop caching prediction matrices, so `summary()` recomputes.

### Also verified

- `posterior_predict` for all four two-block families draws the full mixture including the zeros, so no decomposition is needed anywhere (already covered above).
- On a live `hurdle_gamma` `bnec()` fit, `pp_check`, `pp_check(type = "stat_grouped", stat = "sd")` and `pp_check(type = "loo_pit_overlay")` all return a ggplot, and `bf$criteria` holds `loo,waic` — `bnec()` already attaches the criterion at `R/add_criteria.R`.
- Zero-fraction statistic on the same fit: obs 0.467, simulated 0.495, ppp 0.675. Nothing currently reports this, and it is the statistic these families were added for (#104).

Ready to leave the "needs a decision first" list once (b)'s threshold question above is settled — that is the only open item.

**Comment by beckyfisher** [comment]
## (d) prototyped — and it settles the threshold question

The `bnec_hurdle()` pair finished, so I built the combined simulation rather than leaving it as a claim. It is about six lines:

```r
alive <- posterior_predict(sb)                 # bernoulli, every exposed row
grow  <- posterior_predict(gb, newdata = d)    # full data, not the survivors subset
n     <- min(nrow(alive), nrow(grow))
comb  <- alive[seq_len(n), ] * grow[seq_len(n), ]
```

`newdata = d` rather than the growth fit's own data is the one thing to get right — the growth component is fitted on survivors only (48 of 90 rows here), so it has to be predicted onto the full exposed set for the product to line up with the observed response.

Result is a 400 x 90 matrix matching the observed response:

| | obs | sim | ppp |
|---|---|---|---|
| zero fraction | 0.467 | 0.483 | 0.605 |
| mean | 1.903 | 1.747 | 0.252 |
| sd | 2.301 | 2.270 | 0.44 |

The extrapolation concern already noted in `hurdle_component_preds()` — growth extrapolated past the highest concentration where anything survived — is benign, exactly as that comment predicts. At the top concentration the simulated survival probability came out at 0.167 against 1 survivor of 6 observed, i.e. 0.167. The product is carried by the survival term there, as intended.

### The threshold question now has an answer

I flagged above that the control check had little power at n = 6, on a single `hurdle_gamma` fit. The pair gives an independent second measurement of the same quantity, and it agrees:

| | single `hurdle_gamma` | `bnec_hurdle()` pair |
|---|---|---|
| obs control mean | 4.50 | 4.50 |
| sim control mean | 5.53 | 5.61 |
| `ppp` | 0.82 | ~0.8 |

True value at the control for this simulation is 4.77. So both fits overshoot by roughly 18%, it reproduces across two independent parameterisations, and it is a property of the `ecxexp` shape rather than noise — this is precisely the "model forces the curve away from the control" case that motivated the addition. **Neither `ppp` comes close to flagging it.**

That decides it, as far as I can see: the automatic message in (b) should threshold on the **ratio**, with the `ppp` reported alongside for context. A `ppp` threshold would have stayed silent on both fits, and silence reads as a pass.

Unless you want the other option, this is now fully specified. `notes/tasks/148-model-fit-diagnostics.md` has been updated with the prototype and both measurements.

**Comment by beckyfisher** [comment]
## Checked against the workflow we actually run

Before this gets tackled I went back over the analysis scripts from a recent multi-species project — the real workflow rather than the documented one — to see whether anything there belongs in this issue. Five things do. None of them change a settled decision; three sharpen one, and two are requirements that were not visible from the packaged examples.

### 1. `pp_check.bayesmanecfit()` needs a `model = ` argument, not only the averaged view

The project hand-rolled a per-fit residual check: `DHARMa::createDHARMa()` fed a `posterior_predict()` matrix and a column-mean `posterior_epred()`, with `integerResponse` set from the family, applied one candidate model at a time via `pull_brmsfit()`.

Decision 4 (LOO-PIT, no `DHARMa`) still looks right to me — and note the family-dependent `integerResponse` switch is a piece of bookkeeping LOO-PIT does not need, which is a further small argument for it. But the *access pattern* is the thing to take from this: every check in practice is run **per candidate model**, because that is the level at which a model gets excluded. Decision 2 settled how the *averaged* check is rendered; it did not settle whether the per-model check is reachable. If `pp_check()` on a `bayesmanecfit` only ever returns the averaged view, users go straight back to `pull_brmsfit()` in a loop, which is what the current code does.

So: `pp_check(manec_fit, model = "nec4param")` should delegate to that model's `brmsfit`, with the averaged check as the default when `model` is missing.

### 2. `dispersion()` is used *before* fitting, as a family-selection screen

This is the use I had not accounted for in decision (c). The pattern throughout the scripts is:

```r
tt <- bnec(y | trials(trials) ~ crf(log.x, model = "nec4param"), data = dat)
tt$dispersion   # > 1, so over-dispersed — refit the full set as beta_binomial
```

One cheap single-model fit, read `$dispersion`, and switch `binomial` → `beta_binomial(link = "identity")` if it exceeds 1. That decision was made this way for 7 of 16 datasets and is what the methods section cites as the basis for the family choice.

Two consequences here:

- Decision (c) — keep `dispersion()` — is right, and for a more specific reason than "it's already documented". Its job is a **whole-fit over-dispersion verdict on the two families with no free dispersion parameter**, used to *choose a family*. The docs should say that, because it is not obvious from the current help page.
- `check_fit()` must not be presented as superseding it. The per-group location/scale table cannot answer "binomial or beta-binomial", because once you have fitted `beta_binomial` the free dispersion parameter absorbs the discrepancy — which is exactly the argument in §3 of the scoping comment above, now cutting the other way. Two functions, two questions, and the docs should draw that line explicitly rather than leaving them looking redundant.

This is also the manual form of #149.

### 3. Decision (b): the `summary()` line is the load-bearing half, the message is not

Worth being explicit about, because it affects where the effort goes. Production fitting is batched — dozens of fits dispatched across `future`/`furrr` workers, saved to `.RData`, and the diagnostics rendered from the saved objects some days later. A message emitted at the end of `bnec()` goes to a worker log that nobody reads. It is still worth having for interactive use, but it will be silent in exactly the workflow that most needs the check.

That makes the `summary()` line the part that has to work, and it makes the "do not cache on the fitted object" decision load-bearing rather than incidental: the check has to recompute correctly from an object loaded off disk in a fresh session, with no refit. Worth a test that does precisely that — `save()`, fresh session, `load()`, `summary()`.

### 4. `check_fit()` output has to be batch-usable and non-interactive

Everything in the practical workflow ends up `knitr::kable()`-ed into a per-dataset HTML report, rendered inside a parallel worker with no console. So:

- the return value should be a plain data frame that `rbind`s cleanly across candidate models *and* across fits (the `model` column in the spec covers the first; nothing breaks the second, just worth not regressing);
- the `plot`/`autoplot` method from decision 6 needs a non-interactive path. There are two existing precedents and they differ — `check_priors(ask = FALSE)` and `check_chains(filename = )`. Pick one of them rather than inventing a third.

The report those scripts generate is, in order: `summary()`, model-averaged plot, all-model plot, divergent transitions table, bulk-ESS table, priors table, `check_chains()`, `check_priors(ask = FALSE)`. `check_fit()` slots in after the priors table. Four of those eight are hand-rolled because the package has no accessor.

### 5. Per-candidate-model rows — confirmed

The spec already assumes per-candidate rows as well as an averaged row for a `bayesmanecfit`, on the argument that a global `elpd` can leave a candidate with high weight while it fits the control badly. The practical reports do exactly this for every diagnostic they show, so that assumption holds up against real use.

---

## What is *not* for this issue

The rest of what the scripts expose is sampler-side and workflow-side rather than fit-side — no divergent-transition accessor, no ESS diagnostic, no screening helper, and no documented exclusion step anywhere in the vignettes. That has become its own issue: **#219**, a dedicated end-to-end analysis workflow vignette, which is sequenced to be built *after* this one lands so that `check_fit()` and the `pp_check()` methods can be shown in it. The gaps are listed there.

One item there does touch this issue directly, so flagging it here: `vignettes/example2.Rmd.orig:102` claims "the summary would also include a warning if there were fits with divergent transitions", and `print.manecsummary` warns on Rhat only (`R/print.R:136`, and at a hard-coded 1.05 rather than the 1.01 the same vignette recommends). If decision (b) is adding a line to `summary()` anyway, that is the moment to make the claim true or remove it.

**Comment by beckyfisher** [comment]
## Part D — sampler diagnostics and screening, folded in

The helpers I listed on #219 belong here, not there. #219 is the vignette; these are diagnostic methods, which is what this issue is titled for. Moving them in, and #219 now depends on this issue for all of them rather than listing them as gaps to work around.

The framing that makes this one issue rather than two: `check_fit()` asks *does this model reproduce the data*, and the Part D helpers ask *did this model sample properly*. Both exist to answer the same downstream question — **which candidate models should be in the averaged set** — and both feed the same `amend(drop = )` step and the same `summary()` block. Splitting them would mean designing that step twice.

### D1 — a per-model sampler diagnostic table

The current interface is one model at a time and awkward to read in bulk: `rhat()` returns a nested list, and the documented idiom in `example2` is `rhat(exp_5) |> sapply("[[", "failed")`. Divergences and ESS have no interface at all — the project scripts reach for `rstan::get_num_divergent(pull_brmsfit(fit, model = m)$fit)` and `posterior::summarise_draws()`, per model, by hand.

Proposal: one function returning a data frame, one row per candidate model:

| model | max_rhat | min_ess_ratio | n_divergent | failed |
|---|---|---|---|---|

Name: **`check_sampling()`**, sitting with `check_chains()` (by eye), `check_priors()` (priors) and `check_fit()` (the data). `rhat()` stays as it is — it is documented in `example2` and in the JSS paper.

### D2 — two things I verified, because they change the answer

**Divergences: use `brms::nuts_params()`, not `rstan`.** The project code uses `rstan::get_num_divergent(x$fit)`, and `rstan` is only in `Suggests` here. `brms::nuts_params(x, pars = "divergent__")` is exported, is backend-agnostic, and returns one row per draw per chain, so `sum(np$Value)` is the count. Confirmed on `manec_example`. No new dependency, and it works under the `cmdstanr` backend, which `rstan::get_num_divergent()` only happens to survive.

**ESS: `brms::neff_ratio()` exists and is free — but it is not quite what the project script computes.** This is the one real decision in Part D. Measured on `nec4param` from `manec_example`:

| parameter | `ess_bulk` | `ess_tail` | `neff_ratio` |
|---|---|---|---|
| `b_bot_Intercept` | 10.16 | 34.54 | 0.102 |
| `b_top_Intercept` | 94.78 | 75.57 | 0.756 |
| `b_beta_Intercept` | 13.30 | 44.04 | 0.133 |
| `b_nec_Intercept` | 57.97 | 32.96 | 0.330 |

`ndraws` is 100, so `neff_ratio` is **`min(ess_bulk, ess_tail) / ndraws`** in every row. So:

- **`brms::neff_ratio()`** — no new dependency, one combined number per parameter, and arguably the better screening statistic since it flags a bad tail as well as a bad bulk. Cannot report bulk and tail separately.
- **`posterior::summarise_draws()`** — reports both, which is what the project report tabulates, at the cost of adding `posterior` to `Imports`. Zero install cost in practice (`brms` imports it already), but formally a new dependency, and **the definition of done in the task note currently says "no new package dependency in `DESCRIPTION`"**. So this needs an explicit answer rather than being assumed.

My preference is `neff_ratio()` and one `min_ess_ratio` column — it satisfies the definition of done, and a model that fails on tail ESS should be flagged regardless. But the project screen deliberately looked at bulk specifically, so say if that distinction was doing work.

### D3 — the screening helper

The whole of the project's `clean_models()` is: run the three checks, take the union of failures, `amend(drop = )`. Every project rewrites it. Proposal is table-as-primitive plus a thin wrapper — `check_sampling()` returns the diagnostics, and a screening function applies the thresholds and returns the amended fit, **messaging what it dropped and why**. That message is not decoration: it is what a methods section has to cite.

Two decisions:

**(a) Thresholds.** Rhat 1.01 or 1.05? Note the current position is more tangled than it looks — `rhat()` *defaults* to 1.05 and `print.manecsummary` hard-codes 1.05 with no way to change it (`R/print.R:136`), while `example2` recommends 1.01 as "more conservative" and the project used 1.01 throughout. For divergences the project dropped above **10**, not above 0, because these non-linear models routinely produce a handful at the boundary and a zero threshold would drop nearly everything. That is defensible but it is a judgement call, and whatever default ships will be quoted in methods sections, so it should be a deliberate choice rather than inherited from one project.

**(b) Does the screen include `check_fit()` failures?** I would say **no** — sampler failures are a reason the fit is unusable, whereas a poor `check_fit()` is a modelling result, and dropping on it silently would hide exactly what the user needs to see. Report them side by side, screen on the sampler only.

### D4 — `amend(drop = )` needs a guard

Prerequisite for D3. The union of failure lists can name a model that was never fitted or was already dropped, so the project code has to `intersect(remove_fits, names(fit$mod_fits))` before every call. One line in `amend.bayesmanecfit()`.

### D5 — this merges with decision (b) rather than sitting next to it

Decision (b) already commits to a line in `summary()` for the control fit check. With Part D in the same issue that becomes **one** summary block covering both axes — sampler and fit — rather than two separate additions made independently. It is also the moment to deal with the hard-coded 1.05 in `print.manecsummary`, and to add divergences, which the summary currently never mentions.

Which in turn fixes the doc error: `vignettes/example2.Rmd.orig:102` claims "the summary would also include a warning if there were fits with divergent transitions", and it does not. Part D makes the claim true instead of requiring it to be deleted.

### Sequencing — Part D is the independent half

Worth noting because it changes how this lands. Part D touches no posterior predictive machinery, needs no `loo`, and **does not depend on PR #217** — the sampler diagnostics are per candidate model, so nothing needs a reproducible averaged draw. Parts A and B both do. So D can be built and merged first, on its own, while A and B wait on #217.

That also gets the screening step into the package before #219 is written, which is the sequencing that vignette wants.

### One incidental finding

`manec_example` fails its own default Rhat criterion — both `nec4param` and `ecx4param` return `failed = TRUE` at `rhat_cutoff = 1.05`, and `b_bot_Intercept` has an `ess_bulk` of 10 out of 100 draws. It is a deliberately tiny stored object, so this is not a bug, but it does mean `manec_example` cannot be used to demonstrate screening in an example or a vignette: the screen drops everything and there is no averaged set left. Tests and examples for Part D need a different fixture, and so does #219.

**Comment by beckyfisher** [comment]
## Part D decisions settled

| | |
|---|---|
| D1 | `check_sampling()`, one data frame, one row per candidate model |
| D2 | `brms::neff_ratio()`, one `min_ess_ratio` column. No new dependency, backend-agnostic |
| D3a | Rhat **1.01**, divergences **10**, all thresholds as arguments with those as defaults only |
| D3b | The screen acts on sampler diagnostics only, not on `check_fit()` |
| D4 | Guard lives in the screening function; `amend()` is not touched |
| D5 | One `summary()` block covering sampler and fit |

Three things came out of settling these that change details rather than decisions, plus one loose end that needs a yes/no.

---

### The no-dependency rule was mine, and it is now downgraded

Worth answering directly since you asked where it came from: it is `notes/implementation/00_protocol.md:59`, a working rule written for autonomous implementation sessions so that a session hitting a dependency would **stop and ask** rather than quietly add to `DESCRIPTION`. It was never a package or CRAN constraint, and it should not have been phrased as a hard constraint in the definition of done. Reclassified: a new dependency is allowed, it just has to be a decision on the issue rather than a side effect of an implementation session.

As it happens Part D needs none. Both routes are `brms` exports, and `brms` is in `Depends`.

### Backend agnosticism is satisfied, and this is the reason to avoid the obvious route

`brms::nuts_params(x, pars = "divergent__")` and `brms::neff_ratio(x)` are both generics on `brmsfit` and work under either backend. `rstan::get_num_divergent(x$fit)` — the route the project scripts use — does **not**: it reaches past `brms` into the `stanfit` slot, and `rstan` is only in `Suggests` here. That is a comment worth putting in the code at the point where the choice is made, because `get_num_divergent()` is the obvious thing to reach for and it happens to work often enough to look correct.

### D3a — the reference, and a consequence for the ESS threshold

**Vehtari, A., Gelman, A., Simpson, D., Carpenter, B., & Bürkner, P.-C. (2021). Rank-normalization, folding, and localization: An improved R̂ for assessing convergence of MCMC (with discussion). *Bayesian Analysis*, 16(2), 667–718. doi:10.1214/20-BA1221.**

That is a direct citation for 1.01 — "at convergence, Rhat <= 1.01". It is also already the reference the project's own methods text cites, so it is the right one to put in the docs.

The same paper is where the ESS guidance comes from, and this is the part that needs your call: **its recommendation is an absolute ESS, not a ratio** — greater than 100 per chain, so 400 at the default 4 chains. The 0.1 ratio the project screen used is a Stan/`bayesplot` display convention, not a Vehtari threshold. The two are not interchangeable, and the gap is large in both directions:

| configuration | post-warmup draws | ratio 0.1 means | ESS 400 means |
|---|---|---|---|
| `bayesnec` defaults — 4 chains, `iter` 1e4, `warmup` 8000 (`R/helpers.R:640-650`) | 8000 | ESS 800 | ratio 0.05 |
| the project's config — 3 chains, `iter` 30000, `warmup` 29000, `thin` 3 | ~999 | ESS ~100 | ratio 0.40 |

So a fixed 0.1 ratio is **twice as strict** as the citable threshold under package defaults and **four times more permissive** under the settings the project actually ran. A ratio default has no fixed meaning across configurations; an absolute one does, and it is the quantity that governs Monte Carlo error.

**Proposal, needs a yes/no:** keep `min_ess_ratio` as the reported column as agreed, add an `ndraws` column beside it so the absolute is one multiplication away and a small ratio is interpretable, and **threshold on the absolute** — `ess_cutoff = 400`, citable. If you would rather threshold on the ratio, that is fine, but then the default should be chosen against `bayesnec`'s own 8000 draws rather than inherited from a `bayesplot` plotting convention.

### Divergences — 10 as the default, labelled honestly

There is no literature threshold at 10. Stan's own guidance is that *any* divergent transition indicates the sampler failed to explore the posterior properly and the estimates may be biased. 10 is a pragmatic threshold from practice with these particular non-linear models, which routinely produce a handful near the boundary.

I would keep it at 10 as you have decided, and have the documentation **say that plainly** — a pragmatic loosening of Stan's guidance, not something with support behind it. It will end up quoted in methods sections, and the honest version is more defensible than an implied citation.

Signature, with everything as arguments and `*_cutoff` matching `rhat()`'s existing naming:

```r
check_sampling(x, rhat_cutoff = 1.01, ess_cutoff = 400, divergence_cutoff = 10)
```

with the screening function taking the same four and passing them through.

### D4 — agreed, and the real behaviour is not what the project comment claims

Checked rather than trusted, because the script's comment says `amend()` "errors/warns if asked to drop models that don't exist" and that is not what happens:

| call | result |
|---|---|
| `amend(manec_example, drop = "notamodel")` | two messages, **returns the fit unchanged**. No error |
| `amend(manec_example, drop = c("nec4param", "notamodel"))` | works — drops `nec4param`, returns a `bayesnecfit` |
| `amend(manec_example, drop = c("nec4param", "ecx4param"))` | **error** — "All models removed, nothing to return" |

So the mixed case, which is the common one, already works. `amend()` never errors on an absent name; it silently returns unchanged when *every* named model is absent (`handle_set()` at `R/helpers.R:154` hits `identical(sort(x), sort(tmp))` and returns the `"wrong_model_output"` sentinel).

That supports your call. "Returns unchanged with a message" is a reasonable contract for someone calling `amend()` directly, and only the screening wrapper needs it to be stricter. So `amend()` stays untouched and the wrapper owns three cases, all of which it has to handle anyway:

1. **nothing failed** — don't call `amend()` at all, return unchanged, message that all models passed;
2. **some failed** — `intersect()` against `names(x$mod_fits)`, then `amend(drop = )`;
3. **all failed** — `amend()` errors. The wrapper must catch this and say something more useful than "All models removed, nothing to return", because it is the case where the user most needs to be told what to do next.

Part D therefore touches no existing exported function. Smaller change than scoped.

### CRAN — noted, and the pattern already exists

Examples for `check_sampling()` and the screening function go in `\dontrun{}`, which is already what 22 of the 85 `man/*.Rd` files do, `rhat()` among them. And the #219 vignette is precompiled from `.Rmd.orig`, so it does not run at build or check time either. Nothing in Part D adds to CRAN check time.

The consequence is that the tests carry the verification rather than the examples, and **the tests need a fixture that does not currently exist**. `manec_example` exercises case 3 above perfectly — both its models fail at any cutoff — but nothing in the package exercises case 2, a set where some models pass and some fail, which is the whole point of the feature. A small stored fixture will have to be built. Flagging it now because it is the kind of thing that gets discovered halfway through implementation.

### One loose end

If `summary()` starts reporting at 1.01 under D5, then `rhat()` still *defaulting* to 1.05 is inconsistent, and so is the hard-coded 1.05 in `print.manecsummary` (`R/print.R:136`). Aligning all three on 1.01 is the coherent option and is what the Vehtari citation supports, but it is a user-facing behaviour change to a documented function and needs a `NEWS.md` entry. Say if you would rather leave `rhat()`'s default alone and have only the new functions use 1.01 — that is defensible on backwards-compatibility grounds, it just means the package recommends two different numbers.

**Comment by beckyfisher** [comment]
## ESS and Rhat settled — but D5 is a rewrite, not a number change

Both answered, and the Rhat alignment turned out to be larger than it looked once I traced where `summary()` actually gets its verdict from.

### ESS — report the absolute, threshold at 400

`neff_ratio()` is `min(ess_bulk, ess_tail) / ndraws`, so multiplying it back gives the absolute directly:

```r
min_ess <- min(brms::neff_ratio(x)) * brms::ndraws(x)
```

which is the minimum over parameters of `min(bulk-ESS, tail-ESS)`. Vehtari's recommendation is that **both** bulk-ESS and tail-ESS exceed 100 per chain, so `min_ess > 400` at four chains is exactly that recommendation, with no arithmetic left for the user and still no new dependency.

Columns become `min_ess` and `min_ess_ratio` — both fall out of the same call, and the screen thresholds on `min_ess`. Keeping the ratio alongside is worth the column because it separates "passed 400 because we drew 8000" from "passed 400 efficiently", which is the difference between a fit that is fine and a fit that is expensive.

```r
check_sampling(x, rhat_cutoff = 1.01, ess_cutoff = 400, divergence_cutoff = 10)
```

**One consequence to document rather than discover.** At the project's configuration — 3 chains, `thin = 3`, ~999 retained draws — an ESS of 400 needs a ratio of 0.40, which is a demanding bar for these correlated non-linear parameterisations. Under `bayesnec` defaults (8000 draws) the same 400 is a ratio of 0.05 and is met easily. So adopting the recommendation will fail some fits that the old 0.1 ratio passed, specifically heavily-thinned ones. The correct response there is to **retain more draws, not to lower the cutoff** — thinning discards information and lowers ESS by construction — and that is Vehtari's position too. Worth saying in the help page, because the tempting fix is the wrong one.

### Rhat 1.01 — and `summary()` does not compute Rhat at all

This is the part that changes the shape of D5. `summary()` builds `rhat_issues` at `R/summary.R:146` as `map(x$mod_fits, "fit") |> map(has_r_hat_warnings)`, and `has_r_hat_warnings()` at `R/helpers.R:262` is:

```r
any(grepl("some Rhats are > 1.05", x, fixed = TRUE))
```

It greps **brms's own captured warning text for a literal string**. So the summary's threshold was never `bayesnec`'s to set — it is brms's, discovered by string match — and "align it to 1.01" is not a number change, it is replacing the mechanism with a real `rhat(x, rhat_cutoff = )` call.

That is a good change independent of the threshold. `brms (>= 2.23.0)` in `DESCRIPTION` is a floor, not a ceiling: if brms rewords or retunes that warning, `has_r_hat_warnings()` returns `FALSE` for every model, silently, with no error, and the summary simply stops warning. I checked that it currently works — on `manec_example` both models come back `TRUE` and the warning does fire — so this is latent fragility rather than a live bug. But it is precisely the failure mode where **silence reads as a pass**, which is the same argument that settled the threshold question for the `bnec()` message earlier in this issue.

Touchpoints for the alignment:

| | |
|---|---|
| `R/rhat.R:41`, `R/rhat.R:62` | `rhat.bayesnecfit` / `rhat.bayesmanecfit` defaults 1.05 → 1.01 |
| `R/bayesnechurdlefit-methods.R:470` | `rhat.bayesnechurdlefit` default, same |
| `R/summary.R:146` | replace `has_r_hat_warnings` with a computed `rhat()` verdict |
| `R/print.R:136` | the message hard-codes "Rhats > 1.05"; it must report the cutoff in use |
| `R/helpers.R:262` | `has_r_hat_warnings()` becomes unused — `summary.R` is its only caller, so delete it |
| `man/rhat.Rd`, `man/rhat.bayesnechurdlefit.Rd` | regenerate |
| `NEWS.md` | user-facing behaviour change, needs an entry |

Tests are low risk: `test-bayesmanec_methods.R:32-36` calls `rhat(manec_example)` at the default and at `rhat_cutoff = 1`, and `manec_example` fails at both 1.05 and 1.01, so no expectation changes.

### `example2` needs a rewrite in the same PR

The passage at `example2.Rmd.orig:146-158` exists *entirely* to contrast a permissive default against a conservative override, and the rendered output shows the contrast working:

```
rhat(exp_5)                       # all 8 models FALSE
rhat(exp_5, rhat_cutoff = 1.01)   # ecx4param TRUE
```

With the default at 1.01 those two chunks become **identical** and the prose around them is wrong. Rewrite so the default is 1.01, cite Vehtari for it, and — if a contrast is still wanted — contrast against the looser 1.05 as the historical value.

Worth fixing the same sentence while it is open: "Here we get a message because none of our models failed the default Rhat criterion" has it backwards. No message appears in the rendered output, and `rhat.bayesmanecfit` only messages when **all** models fail (`R/rhat.R:69-71`).

**Gating.** The `.Rmd.orig` edit lands with the PR, but the rendered `.Rmd` stays stale until the #190 precompile run, so the published vignette will show the old prose against the new default in the interim. That needs to be called out in the PR body rather than left to be noticed.

---

Nothing open on Part D now. Task note updated.

**Comment by beckyfisher** [comment]
Complete on `dev`. Closing manually — both PRs targeted `dev`, not the default
branch, so GitHub did not auto-close on merge.

Delivered by two PRs, against the spec in `notes/tasks/148-model-fit-diagnostics.md`:

- **#226** — Parts A, B and C, plus decisions (b) and (d). `check_fit()` reports,
  per group of the predictor, the observed location and scale of the response
  against what the fitted model simulates, with a posterior predictive p-value for
  each and the control flagged; `pp_check()` methods for all three fit classes,
  including `type = "loo_pit_overlay"` (which is what #56 became); (d) the combined
  hurdle check; and (b) the automatic surfacing of a control lack-of-fit, once at
  the end of `bnec()` and as a line in `summary()`. Both threshold on the **ratio**
  of observed to simulated, not on the `ppp` — `fit_ratio_cutoff = 1.15`, confirmed
  by RF 2026-08-24. The measured reason: on two independently fitted
  parameterisations of the same data the simulated control mean overshot the
  observed by ~19% while both p-values sat at ~0.82.
- **#240** — Part D. `check_sampling()` (one row per candidate model: max Rhat, min
  ESS and ESS ratio, divergences, a `failed` verdict), `screen_models()`, and the
  Rhat default aligned on 1.01 after Vehtari et al. (2021). Divergences come from
  `brms::nuts_params()` rather than `rstan::get_num_divergent(x$fit)`.

Verified on `dev` at `879ab53f`: `check_fit`, `check_sampling`, `screen_models` and
the three `pp_check` methods are all exported in `NAMESPACE`.

Two things that rode on this issue and are tracked elsewhere:

- The `example2` doc fix is in `example2.Rmd.orig` (the 1.01 default is documented
  there now). The **rendered** vignette still shows 1.05 — that is #248, under the
  release precompile #190.
- The workflow vignette that depends on Part D is PR #238 (#219), still open. It
  hand-rolled a screen because Part D was thought absent; it can now call
  `screen_models()`.
